File: ui/src/app/model/customer.ts
import { Entitlement } from "./entitlement";



export class Customer {

  id: number;
  name: string;
  userId: string;
  password: string;
  clientType: string;
  priority: number;
  entitlements: Entitlement[];
  disabled: boolean;
  static readonly clientTypes = {
    'c': 'Client',
    'r': 'Reporter',
    'a': 'Administrator'
  }

  constructor(id: number = null, name: string = null, 
    userId: string = null, password: string = null, 
    clientType: string = null, priority: number = null,
    entitlements: Entitlement[] = null, disabled: boolean = false) {
    this.id = id;
    this.name = name;
    this.userId = userId;
    this.password = password;
    this.clientType = clientType;
    this.priority = priority;
    this.entitlements = entitlements;
    this.disabled = disabled;
  }
  
  static copy(customer: Customer, {
      id = customer.getId(),
      name = customer.getName(),
      userId = customer.getUserId(),
      password = customer.getPassword(),
      clientType = customer.getClientType(),
      priority = customer.getPriority(),
      entitlements = customer.getEntitlements(),
      disabled = customer.isDisabled()
    }: {
      id?: number,
      name?: string,
      userId?: string,
      password?: string,
      clientType?: string,
      priority?: number,
      entitlements?: Entitlement[],
      disabled?: boolean
    } = {}): Customer {
      return new Customer(id, name, userId, password, 
        clientType, priority, entitlements, disabled);
  }

  getId(): number {
    return this.id;
  }
  getName(): string {
    return this.name;
  }
  getUserId(): string {
    return this.userId;
  }
  getPassword(): string {
    return this.password;
  }
  getClientType(): string {
    return this.clientType;
  }
  clientTypeString(): string {
    return Customer.clientTypes[this.clientType];
  }
  getPriority(): number { 
    return this.priority;
  }
  getEntitlements(): Entitlement[] {
    return this.entitlements;
  }
  setEntitlements(entitlements: Entitlement[]) {
    this.entitlements = entitlements;
  }
  isDisabled(): boolean {
    return this.disabled;
  }

  equals(other: Customer): boolean {
    return this.getId() === other.getId() &&
    this.getName() === other.getName() &&
    this.getUserId() === other.getUserId() &&
    this.getPassword() === other.getPassword() &&
    this.getClientType() === other.getClientType() &&
    this.getPriority() === other.getPriority();
  }
}

File: src/main/java/com/mni/model/customer/Customer.java
package com.mni.model.customer;

import com.mni.model.entitlement.Entitlement;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

/**
 * Created by will.schick on 6/17/19.
 */
@Entity
public class Customer {

    public static final int MAX_NAME_LENGTH = 100;
    public static final int MAX_USERID_LENGTH = 20;
    public static final int MIN_PASSWORD_LENGTH = 15;
    public static final int MAX_PASSWORD_LENGTH = 100;

    @Id
    @GeneratedValue
    private Long id;

    @Column(unique=true, length=MAX_NAME_LENGTH)
    private String name;

    @Column(unique=true, length=MAX_USERID_LENGTH)
    private String userId;

    @Column(length=MAX_PASSWORD_LENGTH)
    private String password;

    private Character clientType;

    private Integer priority;

    @NotNull
    private boolean disabled;

    @OneToMany(orphanRemoval = true, cascade = CascadeType.ALL, mappedBy = "client")
    @NotNull
    Collection<Entitlement> entitlements = new ArrayList();

    public Customer() {}

    public Customer(Long id, String name, String userId, String password,
                    Character clientType, Integer priority,
                    Collection<Entitlement> entitlements) {
        this.id = id;
        this.name = name;
        this.userId = userId;
        this.password = password;
        this.clientType = clientType;
        this.priority = priority;
        this.entitlements = entitlements;
    }

    public Customer(Long id, String name, String userId, String password,
                    Character clientType, Integer priority,
                    Collection<Entitlement> entitlements, boolean disabled) {
        this(id, name, userId, password, clientType, priority, entitlements);
        this.disabled = disabled;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public Collection<Entitlement> getEntitlements() {
        return entitlements;
    }

    public void setEntitlements(Collection<Entitlement> entitlements) {
        if(this.entitlements == null) this.entitlements = new ArrayList();
        this.entitlements.clear();
        this.entitlements.addAll(entitlements);
    }

    public void setDisabled(boolean disabled) { this.disabled = disabled; }

    public boolean isDisabled() { return disabled; }

    public Character getClientType() {
        return clientType;
    }

    public void setClientType(Character clientType) {
        this.clientType = clientType;
    }

    public Integer getPriority() {
        return priority;
    }

    public void setPriority(Integer priority) {
        this.priority = priority;
    }

    @Override
    public String toString() {
        char[] chars = new char[password.length()];
        Arrays.fill(chars, '*');
        String asterisks = new String(chars);
        return "Customer{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", userId='" + userId + '\'' +
                ", password='" + asterisks + '\'' +
                ", clientType=" + clientType +
                ", priority=" + priority +
                ", disabled=" + disabled +
                ", entitlements=" + entitlements +
                '}';
    }
}
